### PR TITLE
added class MicrotiterPlate

### PR DESCRIPTION
**Pull request by dvnstrcklnd**
- added class MicrotiterPlate

- added Microtiter Plate Test protocol

currently integrates `Collection` and `PlateLayoutGenerator` to yield plate indices in the order specified by the `PlateLayoutGenerator` but only if they are not already filled in the `Collection`.

### CODE REVIEW

**Comment by dvnstrcklnd** [comment]
Note that I ended up using low-level Collection methods in setting up the plate for the test protocol. It would be good to work out how to do this using Association Management or Collection Management modules. I just couldn't get them to work.

**Comment by dvnstrcklnd** [comment]
@malloc3  I took out the stripwell rack class that you had started for now. We should add it back in once we have clarity on the plate class. They should be very similar, and probably inherit the same abstract class

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    mtch = alphanum.match(/\d+/)
+    return mtch[0].to_i if mtch
+  end
+end
+
+# Factory class for building `MicrotiterPlate`s
+# @author Devin Strickland <strcklnd@uw.edu>
+#
+class MicrotiterPlateFactory
+  # Builds a new `MicrotiterPlate`
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb:55]: punctuation?

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  #
+  # @param collection [Collection] the `Collection` that is to be managed
+  # @param data_matrix [String] the data matrix layer that is indicates whether
+  #   the collection has been filled
+  # @param group_size [Fixnum] the size of groups of wells, e.g., correspinding
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb]: spelling: correspinding

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      method: method
+    )
+  end
+end
+
+class MicrotiterPlate
+  # Instantiates `MicrotiterPlate`
+  #
+  # @param collection [Collection] the `Collection` that is to be managed
+  # @param data_matrix [String] the data matrix layer that is indicates whether
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb]: you need to explain "filled".  how is it different than having a part in the collection?

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  #   the collection has been filled
+  # @param group_size [Fixnum] the size of groups of wells, e.g., correspinding
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb]: correspinding

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  def initialize(collection:, data_matrix:, group_size:, method:)
+    @collection = collection
+    @data_matrix = data_matrix
+    @plate_layout_generator = PlateLayoutGeneratorFactory.build(
+      group_size: group_size,
+      method: method
+    )
+  end
+
+  # Returns the next `PlateLayoutGenerator` index that does not point to a
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb:121]: is it the next "unused" index?

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  #   slot that is already filled in the `data_matrix` layer of
+  #   the `Collection`
+  #
+  # @param data_matrix [String] label of a data matrix for the collection that
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb]: no such parameter

File: microtiter_plates/libraries/microtiterplates/source.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+# frozen_string_literal: true
+
+needs 'Microtiter Plates/PlateLayoutGenerator'
+
+module MicrotiterPlates
+  # Convert a letter to the corresponding array index
+  #
+  # @param letter [String] the letter (usually of a row)
+  # @return Fixnum
+  def letter_to_index(letter)
```
Comment (bjkeller) [inline microtiter_plates/libraries/microtiterplates/source.rb:11]: how many implementations of alphanumeric coordinates do we have?

**Review by bjkeller — COMMENTED** [review]
I'm not sure what the point of the data matrix is and the comments aren't helping.

Are these the only methods that are needed for a microtiter plate?

File: microtiter_plates/libraries/microtiterplates/source.rb (hunk 1 above)
Comment (dvnstrcklnd) [reply microtiter_plates/libraries/microtiterplates/source.rb:55]: You mean it needs a period?

File: microtiter_plates/libraries/microtiterplates/source.rb (hunk 3 above)
Comment (dvnstrcklnd) [reply microtiter_plates/libraries/microtiterplates/source.rb]: `data_matrix` is probably a misunderstanding from the AssociationManagement module. I was envisioning parallel matrices attached to the collection. Her it is referring to whether the DataAssociation for a part has something at the key `data_matrix`.

Comment (dvnstrcklnd) [reply microtiter_plates/libraries/microtiterplates/source.rb]: Happy to rename this param.

File: microtiter_plates/libraries/microtiterplates/source.rb (hunk 5 above)
Comment (dvnstrcklnd) [reply microtiter_plates/libraries/microtiterplates/source.rb:121]: Yes.

**Comment by dvnstrcklnd** [comment]
These probably are not the only methods that are needed, but are enough for what I am needing to do.

**Pull request merged**